internal/migration/migration.go

```diff
@@ -0,0 +1,406 @@
+	"fmt"
+	"os"
+	"time"
+
+	json "github.com/nikkolasg/hexjson"
+	"go.etcd.io/bbolt"
+
+	"github.com/drand/drand/chain"
+	"github.com/drand/drand/chain/postgresdb/database"
+	"github.com/drand/drand/chain/postgresdb/pgdb"
+	"github.com/drand/drand/chain/postgresdb/schema"
+	"github.com/drand/drand/log"
+)
+
+type (
+	beacon struct {
+		PreviousSig []byte `json:",omitempty"`
+		Round       uint64
+		Signature   []byte
+	}
+
+	migrator struct {
+		startedAt time.Time
+		logger    log.Logger
+
+		pgDSN            string
+		bufferSize       int
+		beaconName       string
+		sourceBeaconPath string
+
+		existingRows int
+		migratedRows int
+		destination  chain.StorageType
+	}
+)
+
+const (
+	ownerOnly = 0600
+
+	DefaultBufferSize = 10_000
+)
+
+var (
+	// ErrMigrationNotNeeded is returned if the database format is already at the target version
+	ErrMigrationNotNeeded = errors.New("migration not needed")
+
+	bucketName = []byte("beacons")
+)
+
+func Migrate(logger log.Logger, sourceBeaconPath, beaconName string, destination chain.StorageType, pgDSN string, bufferSize int) error {
+	startedAt := time.Now()
+
+	if err := shouldMigrate(logger, sourceBeaconPath, beaconName, destination, pgDSN); err != nil {
+		if errors.Is(err, ErrMigrationNotNeeded) {
+			logger.Warnw("decided storage format migration is not needed", "err", err)
+		}
+		return err
+	}
+
+	bufferSize, err := computerBufferSize(bufferSize, logger, sourceBeaconPath)
+	if err != nil {
+		return err
+	}
+
+	m := migrator{
+		logger:    logger,
+		startedAt: startedAt,
+
+		pgDSN:            pgDSN,
+		bufferSize:       bufferSize,
+		beaconName:       beaconName,
+		sourceBeaconPath: sourceBeaconPath,
+		destination:      destination,
+	}
+
+	return m.doMigrate()
+}
+
+func computerBufferSize(bufferSize int, logger log.Logger, sourceBeaconPath string) (int, error) {
+	if bufferSize < 0 {
+		logger.Infow("buffer size not specified, defaulting to 10000")
+		return DefaultBufferSize, nil
+	}
+
+	if bufferSize == 0 {
+		return automaticBufferSize(logger, sourceBeaconPath)
+	}
+
+	if bufferSize < DefaultBufferSize {
+		logger.Warnw("buffer size seems a bit too small. The migration process might be slow", "bufferSize", bufferSize)
+	}
+
+	//nolint:gomnd // See below
+	if bufferSize > 10_000_000 {
+		//nolint:lll // This line has the right amount of chars
+		logger.Warnw("buffer size seems a bit too large. Make sure your system can allocate enough system memory for this", "bufferSize", bufferSize)
+	}
+	return bufferSize, nil
+}
+
+func automaticBufferSize(logger log.Logger, sourceBeaconPath string) (int, error) {
+	logger.Warnw("buffer size set to 0. Running automatic buffer inference. Make sure you have enough system memory for this")
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return -1, err
+	}
+	defer existingDB.Close()
+
+	var bufferSize = 0
+
+	err = existingDB.View(func(tx *bbolt.Tx) error {
+		bufferSize = tx.Bucket(bucketName).Stats().KeyN
```

Comment:
This feels a little bit extreme xD
How much memory does that typically take? 
For 2'000'000 entries, that would be like 2GB I guess, so looks fine. Or did you test it and it was way more memory?

Reply:
This is going to be a bit more than 2GB because it contains the extra data structures required by BoltDB. Based on the testing I had it's about 4GB for 2'000'000 entries.